main.cpp

```diff
@@ -41,12 +41,207 @@
+        }
+
+        std::memcpy(fuzz_data, encoded_data.data(), encoded_size);
+
+        return encoded_size;
+    }
+    catch (const std::exception &e)
+    {
+        //std::cout << "Exception: " << e.what() << std::endl;
+        // In case of any exception, fall back to initial input
+        return initial_input(fuzz_data, max_size);
+    }
+}
+
+/* A custom cross-over mutator that decodes the bech32 inputs before cross-over
+ * mutating them. Like LLVMFuzzerCustomMutator, this enables more efficient
+ * fuzzing of bolt12 offers, invoice requests, and invoices. */
+extern "C" size_t LLVMFuzzerCustomCrossOver(const uint8_t *in1, size_t in1_size,
+                                            const uint8_t *in2, size_t in2_size,
+                                            uint8_t *out, size_t max_out_size,
+                                            unsigned seed)
+{
+    try {
+        // Decode first input
+        std::string input1_str(reinterpret_cast<const char*>(in1), in1_size);
+        
+        // Check if input1 contains the correct HRP
+        if (input1_str.length() < bech32_hrp.length() + 1 ||
+            input1_str.substr(0, bech32_hrp.length()) != bech32_hrp ||
+            (input1_str.length() > bech32_hrp.length() && 
+             input1_str[bech32_hrp.length()] != '1')) {
+            return 0;
+        }
+
+        bech32::DecodeResult decoded_result1 = bech32::DecodeNoChecksum(input1_str, 
+                                                    bech32::CharLimit::CUSTOM_MUTATOR);
+        std::vector<uint8_t> decoded_data1 = decoded_result1.data;
+        
+        if (decoded_data1.empty()) {
+            return 0;
+        }
+
+        // Decode second input
+        std::string input2_str(reinterpret_cast<const char*>(in2), in2_size);
+        
+        // Check if input2 contains the correct HRP
+        if (input2_str.length() < bech32_hrp.length() + 1 ||
+            input2_str.substr(0, bech32_hrp.length()) != bech32_hrp ||
+            (input2_str.length() > bech32_hrp.length() && 
+             input2_str[bech32_hrp.length()] != '1')) {
+            // std::cout << "// Check if input2 contains the correct HRP";
+            return 0;
+        }
+
+        bech32::DecodeResult decoded_result2 = bech32::DecodeNoChecksum(input2_str, 
+                                                    bech32::CharLimit::CUSTOM_MUTATOR);
+        std::vector<uint8_t> decoded_data2 = decoded_result2.data;
+        
+        if (decoded_data2.empty()) {
+            // std::cout << "decoded_data2.empty()";
+            return 0;
+        }
+
+        // Perform cross-over on decoded data
+        std::vector<uint8_t> crossed_data = cross_over(decoded_data1, decoded_data2, 
+                                                       max_out_size, seed);
+        
+        if (crossed_data.empty()) {
+            // std::cout << "crossed_data.empty()";
+            return 0;
+        }
```

Comment:
Is this check necessary?  If the mutated data is empty, shouldn't the rest of the mutator still work properly?